### PR TITLE
feat: use new push0 opcode

### PR DESCRIPTION
**Pull request by charles-cooper**
per shanghai fork

### What I did
implement #3360 

### Commit message

```
per the shanghai fork - during codegen, use the new `push0` opcode
instead of the `push1 0` sequence
```

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/ir/compile_ir.py [hunk 1]
```diff
@@ -554,36 +555,22 @@ def _height_of(witharg):
...
-                MemoryPositions.FREE_VAR_SPACE,
-                "MSTORE",
-                "PUSH1",
-                32,
-                "PUSH1",
-                MemoryPositions.FREE_VAR_SPACE,
-                "SHA3",
-            ]
+            PUSH(MemoryPositions.FREE_VAR_SPACE) + ["MSTORE"] + PUSH(32),
+            +PUSH(MemoryPositions.FREE_VAR_SPACE) + ["SHA3"],
```
Comment (fubuloubu) [inline vyper/ir/compile_ir.py:559]:
```suggestion
            [
                *PUSH(MemoryPositions.FREE_VAR_SPACE),
                "MSTORE",
                *PUSH(32),
                *PUSH(MemoryPositions.FREE_VAR_SPACE),
                "SHA3",
            ]
```

File: vyper/ir/compile_ir.py [hunk 2]
```diff
@@ -554,36 +555,22 @@ def _height_of(witharg):
...
-                MemoryPositions.FREE_VAR_SPACE,
-                "SHA3",
-            ]
+            PUSH(MemoryPositions.FREE_VAR_SPACE2)
+            + ["MSTORE"]
+            + PUSH(MemoryPositions.FREE_VAR_SPACE)
+            + ["MSTORE"]
+            + PUSH(64)
+            + PUSH(MemoryPositions.FREE_VAR_SPACE)
+            + ["SHA3"]
```
Comment (fubuloubu) [inline vyper/ir/compile_ir.py:573]:
```suggestion
            [
                *PUSH(MemoryPositions.FREE_VAR_SPACE2),
                "MSTORE",
                *PUSH(MemoryPositions.FREE_VAR_SPACE),
                "MSTORE",
                *PUSH(64),
                *PUSH(MemoryPositions.FREE_VAR_SPACE),
                "SHA3",
            ]
```

File: vyper/ir/compile_ir.py [hunk 3]
```diff
@@ -555,13 +556,10 @@ def _height_of(witharg):
         o = _compile_to_assembly(code.args[0], withargs, existing_labels, break_dest, height)
         o.extend(
             [
-                "PUSH1",
-                MemoryPositions.FREE_VAR_SPACE,
```
Comment (fubuloubu) [inline vyper/ir/compile_ir.py:559]: Is this only 0 or 1?

File: vyper/ir/compile_ir.py [hunk 4]
```diff
@@ -23,7 +23,8 @@ def num_to_bytearray(x):
 
 def PUSH(x):
```
Comment (fubuloubu) [inline vyper/ir/compile_ir.py:24]: wonder if this might be better named `push_stack` or something clearer, I thought it was an opcode at first

**Review by fubuloubu — APPROVED** [review]

File: vyper/ir/compile_ir.py (hunk 4 above)
Comment (charles-cooper) [reply vyper/ir/compile_ir.py:24]: could be -- it is supposed to be an abstraction of the PUSH opcodes

File: vyper/ir/compile_ir.py (hunk 3 above)
Comment (charles-cooper) [reply vyper/ir/compile_ir.py:559]: right now FREE_VAR_SPACE is 0, but since it could change, use the PUSH abstraction

**Comment by charles-cooper** [comment]
this is waiting on upstream -- py-evm 0.7.0a1 was released with shanghai updates including PUSH0, but a new eth-tester has not been released which results in a dependency conflict when trying to install the new py-evm.

**Comment by charles-cooper** [comment]
looks like eth-tester updated on may 12, see https://pypi.org/project/eth-tester/#history

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**